File: lua/adminLoginDefense.lua
local failKey = KEYS[1]
local lockKey = KEYS[2]

local action = ARGV[1] or ''

local function to_number(value, fallback)
  local n = tonumber(value)
  if n == nil then
    return fallback
  end
  return n
end

local function to_response(locked, remainingMs, failures)
  return { tostring(locked and 1 or 0), tostring(remainingMs or 0), tostring(failures or 0) }
end

if action == 'ttl' then
  local ttl = redis.call('pttl', lockKey)
  if type(ttl) ~= 'number' or ttl <= 0 then
    ttl = 0
  end
  return tostring(ttl)
end

if action == 'reset' then
  redis.call('del', failKey, lockKey)
  return '1'
end

if action ~= 'fail' then
  error('unsupported admin login defense action: ' .. action)
end

local threshold = to_number(ARGV[2], 5)
local failureWindowMs = to_number(ARGV[3], 15 * 60 * 1000)
local lockoutMs = to_number(ARGV[4], 10 * 60 * 1000)
local maxLockoutMs = to_number(ARGV[5], 60 * 60 * 1000)

local existingLockMs = redis.call('pttl', lockKey)
if type(existingLockMs) ~= 'number' then
  existingLockMs = 0
end
if existingLockMs > 0 then
  return to_response(true, existingLockMs, 0)
end

local count = redis.call('incr', failKey)
if count == 1 then
  redis.call('pexpire', failKey, failureWindowMs)
end

if count >= threshold then
  local nextLockMs = lockoutMs
  if nextLockMs > maxLockoutMs then
    nextLockMs = maxLockoutMs
  end
  redis.call('psetex', lockKey, nextLockMs, tostring(count))
  redis.call('del', failKey)
  return to_response(true, nextLockMs, count)
end

return to_response(false, 0, count)

File: lib/adminLoginDefense.js
'use strict';

const fs = require('fs');
const path = require('path');
const createRedisLuaScript = require('./redisLuaScript');

function toPositiveInteger(value, fallback) {
  const parsed = Number(value);
  if (!Number.isInteger(parsed) || parsed <= 0) {
    return fallback;
  }
  return parsed;
}

function normalizeScope(value) {
  return typeof value === 'string' ? value.trim() : '';
}

function createScopeKeys(KEYS, scope) {
  const normalized = normalizeScope(scope);
  if (!normalized) {
    return null;
  }

  return {
    fail: KEYS.rateAdminLoginFailures(normalized),
    lock: KEYS.rateAdminLoginLock(normalized),
  };
}

function parseInteger(value, fallback = 0) {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function createAdminLoginDefense(redisClient, KEYS, options = {}) {
  if (!redisClient) {
    throw new Error('redisClient is required');
  }
  if (!KEYS) {
    throw new Error('KEYS is required');
  }

  const failureWindowMs = toPositiveInteger(options.failureWindowMs, 15 * 60 * 1000);
  const lockoutMs = toPositiveInteger(options.lockoutMs, 10 * 60 * 1000);
  const threshold = toPositiveInteger(options.threshold, 5);
  const maxLockoutMs = toPositiveInteger(options.maxLockoutMs, 60 * 60 * 1000);

  const luaPath = path.join(__dirname, '..', 'lua', 'adminLoginDefense.lua');
  const luaSource = fs.readFileSync(luaPath, 'utf8');
  const script = createRedisLuaScript(redisClient, luaSource);

  async function getLockRemainingMs(scope) {
    const keys = createScopeKeys(KEYS, scope);
    if (!keys) {
      return 0;
    }

    try {
      const ttl = await script.eval(2, keys.fail, keys.lock, 'ttl');
      return Math.max(0, parseInteger(ttl, 0));
    } catch (err) {
      console.error('admin login defense: lock check failed', err);
      return 0;
    }
  }

  async function recordFailure(scope) {
    const keys = createScopeKeys(KEYS, scope);
    if (!keys) {
      return { locked: false, remainingMs: 0, failures: 0 };
    }

    try {
      const response = await script.eval(
        2,
        keys.fail,
        keys.lock,
        'fail',
        String(threshold),
        String(failureWindowMs),
        String(lockoutMs),
        String(maxLockoutMs)
      );

      if (!Array.isArray(response) || response.length < 3) {
        throw new Error(`Invalid admin login defense response: ${JSON.stringify(response)}`);
      }

      return {
        locked: response[0] === '1',
        remainingMs: Math.max(0, parseInteger(response[1], 0)),
        failures: Math.max(0, parseInteger(response[2], 0)),
      };
    } catch (err) {
      console.error('admin login defense: record failure failed', err);
      return { locked: true, remainingMs: lockoutMs, failures: 0 };
    }
  }

  async function reset(scope) {
    const keys = createScopeKeys(KEYS, scope);
    if (!keys) {
      return false;
    }

    try {
      const result = await script.eval(2, keys.fail, keys.lock, 'reset');
      return result === '1';
    } catch (err) {
      console.error('admin login defense: reset failed', err);
      return false;
    }
  }

  return {
    getLockRemainingMs,
    recordFailure,
    reset,
  };
}

module.exports = createAdminLoginDefense;
